### DISCUSSION TITLE
Possible to export immediately after mark as watch/unwatched?

### DISCUSSION THREAD

**Discussion by namankanakiya** [arabcoders/watchstate#564] [General] [answered]
Hey! First of all, thanks for creating this, it's been super helpful already. I have a setup with webhooks for Emby/Jellyfin, which seems to pick up played / unplayed status pretty much immediately (nice!). I then have the CRON tasks for importing (for newly added items per FAQ) and exporting (to actually update Jellyfin and Emby with the status) running every so often.

My question is whether it's possible to trigger the export task every time there has been a confirmed update to the WatchState internal db - that is, if a webhook from Emby says I've played item A, WatchState accepts and updates item A to be played, can we trigger an export immediately to update Jellyfin as opposed to waiting for the CRON export task.

Thanks! And sorry if this has been asked already, I couldn't find anything similar.

**Comment by arabcoders** [comment]
> Hey! First of all, thanks for creating this, it's been super helpful already. I have a setup with webhooks for Emby/Jellyfin, which seems to pick up played / unplayed status pretty much immediately (nice!). I then have the CRON tasks for importing (for newly added items per FAQ) and exporting (to actually update Jellyfin and Emby with the status) running every so often.
> 
> My question is whether it's possible to trigger the export task every time there has been a confirmed update to the WatchState internal db - that is, if a webhook from Emby says I've played item A, WatchState accepts and updates item A to be played, can we trigger an export immediately to update Jellyfin as opposed to waiting for the CRON export task.
> 
> Thanks! And sorry if this has been asked already, I couldn't find anything similar.

Hi, of coarse you are able to do whatever you want including changing how often the export run, however keep in mind, import and export commands are resource intensive and likely to lead to slow down of your media players as well. We suggest using webhooks for those quick to update tasks like watched/unwatched and rely on import/export to pick up missed ones.

Webhooks should update within 2-3mins max.

**Reply by namankanakiya** [reply]
Sorry, should have been more explicit - I meant using the database change as a trigger for export, vs making the export job run more frequently via a CRON timer (what I think you are suggesting)

**Reply by arabcoders** [reply]
This is already the case with the webhook implementation :-) Export is for the things that the webhook was unable to sync. I hope i understood your problem

**Reply by namankanakiya** [reply]
Thanks for the confirmation of expected behavior - in that case, is there a setting or something to set to get that behavior? What I'm observing is this:
1) at 2:05, I finish playing an episode on Emby
2) Webhook gets received immediately (looking at webhook.date.log)
3) After a minute or two, the database state for that episode is correct for Emby, but NOT for Jellyfin (checked via the History GUI, and clicking on the episode). In Jellyfin itself, the played state of the episode also isn't changed

.... lots of time until CRON timer

4) My CRON timer for export happens - export occurs
5) Jellyfin is updated, both in itself, and in the watchstate db (looking via History -> Episode in the GUI)

So until the CRON export job happens, Jellyfin isn't updated essentially.

After some looking I found `WS_PUSH_ENABLED` in your old news section - is that what it does?

**Reply by arabcoders** [reply]
`WS_PUSH_ENABLED` is gone and we use event system now, you can reduce the item for when tasks run via the env page, look for `WS_CRON_EXPORT_AT` & `WS_CRON_IMPORT_AT`. The webhook should pick the majority of changes unless you have problem with metadata not being added to the other server before the watch event is sent. you can see if that the case via logs mostly the webhook or via the tasks logs

**Reply by joggs** [reply]
I was looking for the exact same thing.  The problem is that CRON is a schedule that we COMMAND to run at an interval. What would be nice is for the watchstate app to have an event listener that listens for received webhook calls and IMMEDIATELY perform the export action, and sends the changes to the other backends. We do not want to wait for a cron event to trigger and it is not healthy to set the cron event to trigger every 10 seconds or so, thats why event listeners are so much better as they immediately react when something happens. The schedule is a greate complement, but I would suggest that you can enable immediate sync for each backend entry . Maybe that was how it worked when you had WS_PUSH? was there any problems that made you remove it?

**Reply by arabcoders** [reply, marked as answer]
Sadly, it's not possible the WS_PUSH was same as the new events system infect it's based on it, we didn't sync immediately it's unpractical to do so, you will be in a never ending loop as each server will sent a response for the request we sent thus it will even effect players like plex etc with random pause and un-pause.

**Comment by joggs** [comment]
Aha, I see. Thanks for the clarification


/Jörgen Bergström


mån 18 nov. 2024 kl. 18:07 skrev Abdulmohsen ***@***.***>:

**Reply by arabcoders** [reply]
No worries. 2min ought to be enough =)
